File: riveredge-frontend/src/apps/kuaiplm/services/rd-project.ts
/**
 * 研发项目 / NPI 工作台 API
 */

import { apiRequest } from '../../../services/api';
import type { EngineeringLinkType } from './master-data-links';

const BASE = '/apps/kuaiplm/rd-projects';

export type ProjectType = 'RD' | 'DELIVERY';

export interface RdProject {
  id?: number;
  uuid?: string;
  project_code?: string;
  project_name?: string;
  project_type?: ProjectType;
  source_project_id?: number | null;
  source_project_code?: string | null;
  status?: string;
  material_id?: number | null;
  material_code?: string | null;
  material_name?: string | null;
  owner_name?: string | null;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  current_gate_key?: string | null;
  current_gate_name?: string;
  notes?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface RdProjectGate {
  id?: number;
  project_id?: number;
  gate_key?: string;
  gate_name?: string;
  sort_order?: number;
  status?: string;
  planned_date?: string | null;
  actual_date?: string | null;
  reviewer_id?: number | null;
  reviewer_name?: string | null;
  review_notes?: string | null;
  criteria?: string | null;
}

export interface RdProjectTask {
  id?: number;
  project_id?: number;
  gate_id?: number | null;
  parent_task_id?: number | null;
  task_name?: string;
  description?: string | null;
  assignee_name?: string | null;
  assignee_id?: number | null;
  status?: string;
  due_date?: string | null;
  priority?: string | null;
}

export interface RdProjectDeliverable {
  id?: number;
  project_id?: number;
  gate_id?: number | null;
  name?: string;
  description?: string | null;
  deliverable_type?: string | null;
  status?: string;
  file_url?: string | null;
  file_name?: string | null;
}

export interface RdProjectLink {
  id?: number;
  project_id?: number;
  link_type?: EngineeringLinkType | string;
  target_type?: string;
  target_name?: string | null;
  target_code?: string | null;
  target_uuid?: string | null;
  target_id?: number | null;
  material_id?: number | null;
  version?: string | null;
  notes?: string | null;
}

export interface ProjectCollaborationSummary {
  requirement_count?: number;
  design_review_count?: number;
  fmea_count?: number;
}

const WORKBENCH_NESTED_KEYS = new Set([
  'gates',
  'tasks',
  'links',
  'deliverables',
  'related_articles',
  'progress',
  'collaboration',
]);

function normalizeWorkbench(raw: Record<string, unknown>): RdProjectWorkbench {
  if (raw.project && typeof raw.project === 'object') {
    return raw as unknown as RdProjectWorkbench;
  }
  const project: Record<string, unknown> = {};
  const nested: RdProjectWorkbench = {
    gates: [],
    tasks: [],
    links: [],
    deliverables: [],
    related_articles: [],
    progress: 0,
    collaboration: {},
  };
  for (const [key, value] of Object.entries(raw)) {
    if (WORKBENCH_NESTED_KEYS.has(key)) {
      (nested as Record<string, unknown>)[key] = value;
    } else {
      project[key] = value;
    }
  }
  return { project: project as RdProject, ...nested };
}

function normalizeLinkPayload(data: Partial<RdProjectLink> & { link_label?: string; version?: string }) {
  const linkType = data.link_type === 'route' ? 'process_route' : data.link_type;
  const notes = [data.notes, data.version ? `版本: ${data.version}` : null].filter(Boolean).join('; ');
  return {
    link_type: linkType,
    target_type: linkType ?? data.target_type ?? 'other',
    target_id: data.target_id != null ? Number(data.target_id) : undefined,
    target_uuid: data.target_uuid,
    target_code: data.target_code,
    target_name: data.target_name ?? data.link_label,
    notes: notes || undefined,
  };
}

export interface RdProjectWorkbench {
  project?: RdProject;
  gates?: RdProjectGate[];
  tasks?: RdProjectTask[];
  links?: RdProjectLink[];
  deliverables?: RdProjectDeliverable[];
  related_articles?: Array<{
    id: number;
    title?: string;
    space_name?: string;
    updated_at?: string;
  }>;
  progress?: number;
  collaboration?: ProjectCollaborationSummary;
}

export interface RdProjectListParams {
  skip?: number;
  limit?: number;
  status?: string;
  lifecycle_stage?: string;
  keyword?: string;
  project_type?: ProjectType;
}

function unwrapList<T>(res: unknown): { items: T[]; total: number } {
  if (Array.isArray(res)) return { items: res as T[], total: res.length };
  const r = res as Record<string, unknown>;
  const items = (r.items ?? r.data ?? r.results ?? []) as T[];
  const total = Number(r.total ?? (Array.isArray(items) ? items.length : 0));
  return { items: Array.isArray(items) ? items : [], total };
}

export async function listRdProjects(params?: RdProjectListParams) {
  const res = await apiRequest<unknown>(BASE, { method: 'GET', params });
  return unwrapList<RdProject>(res);
}

export async function getRdProject(id: number | string) {
  return apiRequest<RdProject>(`${BASE}/${id}`);
}

export async function getRdProjectWorkbench(id: number | string) {
  const raw = await apiRequest<Record<string, unknown>>(`${BASE}/${id}/workbench`);
  return normalizeWorkbench(raw);
}

export async function createRdProject(data: Partial<RdProject>) {
  return apiRequest<RdProject>(BASE, { method: 'POST', data });
}

export async function updateRdProject(id: number | string, data: Partial<RdProject>) {
  return apiRequest<RdProject>(`${BASE}/${id}`, { method: 'PUT', data });
}

export async function deleteRdProject(id: number | string) {
  return apiRequest<void>(`${BASE}/${id}`, { method: 'DELETE' });
}

export async function spawnDeliveryProject(
  projectId: number | string,
  data?: { project_name?: string; project_code?: string; owner_id?: number; owner_name?: string },
) {
  return apiRequest<RdProject>(`${BASE}/${projectId}/spawn-delivery`, { method: 'POST', data: data ?? {} });
}

export async function pushTrialWorkOrder(
  projectId: number | string,
  data?: { quantity?: number; notes?: string },
) {
  return apiRequest<{ work_order_id?: number; work_order_code?: string }>(
    `${BASE}/${projectId}/push-trial-work-order`,
    { method: 'POST', data: data ?? {} },
  );
}

export async function createRdProjectTask(projectId: number | string, data: Partial<RdProjectTask>) {
  return apiRequest<RdProjectTask>(`${BASE}/${projectId}/tasks`, { method: 'POST', data });
}

export async function updateRdProjectTask(
  projectId: number | string,
  taskId: number | string,
  data: Partial<RdProjectTask>,
) {
  return apiRequest<RdProjectTask>(`${BASE}/${projectId}/tasks/${taskId}`, { method: 'PUT', data });
}

export async function deleteRdProjectTask(projectId: number | string, taskId: number | string) {
  return apiRequest<void>(`${BASE}/${projectId}/tasks/${taskId}`, { method: 'DELETE' });
}

export async function createRdProjectDeliverable(
  projectId: number | string,
  data: Partial<RdProjectDeliverable>,
) {
  return apiRequest<RdProjectDeliverable>(`${BASE}/${projectId}/deliverables`, { method: 'POST', data });
}

export async function updateRdProjectDeliverable(
  projectId: number | string,
  deliverableId: number | string,
  data: Partial<RdProjectDeliverable>,
) {
  return apiRequest<RdProjectDeliverable>(`${BASE}/${projectId}/deliverables/${deliverableId}`, {
    method: 'PUT',
    data,
  });
}

export async function deleteRdProjectDeliverable(
  projectId: number | string,
  deliverableId: number | string,
) {
  return apiRequest<void>(`${BASE}/${projectId}/deliverables/${deliverableId}`, { method: 'DELETE' });
}

export async function createRdProjectLink(
  projectId: number | string,
  data: Partial<RdProjectLink> & { link_label?: string; version?: string },
) {
  return apiRequest<RdProjectLink>(`${BASE}/${projectId}/links`, {
    method: 'POST',
    data: normalizeLinkPayload(data),
  });
}

export async function deleteRdProjectLink(projectId: number | string, linkId: number | string) {
  return apiRequest<void>(`${BASE}/${projectId}/links/${linkId}`, { method: 'DELETE' });
}

export async function updateRdProjectGate(
  projectId: number | string,
  gateId: number | string,
  data: Partial<RdProjectGate>,
) {
  return apiRequest<RdProjectGate>(`${BASE}/${projectId}/gates/${gateId}`, { method: 'PUT', data });
}

export const PROJECT_TYPE_LABELS: Record<ProjectType, string> = {
  RD: '研发项目',
  DELIVERY: '交付项目',
};

export const PROJECT_STATUS_LABELS: Record<string, string> = {
  DRAFT: '草稿',
  IN_PROGRESS: '进行中',
  ON_HOLD: '已暂停',
  COMPLETED: '已结案',
  CANCELLED: '已取消',
};

export function buildProjectStatusValueEnum(): Record<string, { text: string; status?: string }> {
  const antStatus: Record<string, string> = {
    DRAFT: 'Default',
    IN_PROGRESS: 'Processing',
    ON_HOLD: 'Warning',
    COMPLETED: 'Success',
    CANCELLED: 'Default',
  };
  return Object.fromEntries(
    Object.entries(PROJECT_STATUS_LABELS).map(([key, text]) => [
      key,
      { text, status: antStatus[key] ?? 'Default' },
    ]),
  );
}

export const TASK_STATUS_LABELS: Record<string, string> = {
  TODO: '待办',
  IN_PROGRESS: '进行中',
  DONE: '已完成',
  CANCELLED: '已取消',
};

export const DELIVERABLE_STATUS_LABELS: Record<string, string> = {
  PENDING: '待提交',
  SUBMITTED: '已提交',
  APPROVED: '已批准',
  REJECTED: '已驳回',
};

export const GATE_STATUS_LABELS: Record<string, string> = {
  PENDING: '待开始',
  IN_PROGRESS: '进行中',
  PASSED: '已通过',
  FAILED: '未通过',
  SKIPPED: '已跳过',
};

File: riveredge-backend/src/apps/kuaiplm/models/rd_project.py
"""
研发项目 / NPI 模型

Author: RiverEdge Team
Date: 2026-05-28
"""

from tortoise import fields

from core.models.base import BaseModel
from apps.kuaiplm.constants.rd_project import (
    GateMilestoneRole,
    RdDeliverableStatus,
    RdGateStatus,
    RdProjectLinkType,
    RdProjectStatus,
    RdProjectType,
    RdTaskStatus,
)


class RdProject(BaseModel):
    """研发项目"""

    tenant_id = fields.IntField(description="租户ID")
    project_code = fields.CharField(max_length=50, db_index=True, description="项目编码")
    project_name = fields.CharField(max_length=200, description="项目名称")
    description = fields.TextField(null=True, description="项目描述")
    status = fields.CharField(
        max_length=30,
        default=RdProjectStatus.DRAFT.value,
        description="项目状态",
    )
    project_type = fields.CharField(
        max_length=20,
        default=RdProjectType.RD.value,
        description="项目类型 RD | DELIVERY",
    )
    source_project_id = fields.IntField(null=True, description="来源研发项目ID")
    gate_template_id = fields.IntField(null=True, description="阶段门模板ID")
    material_id = fields.IntField(null=True, description="目标物料ID")
    material_code = fields.CharField(max_length=50, null=True, description="目标物料编码")
    material_name = fields.CharField(max_length=200, null=True, description="目标物料名称")
    current_gate_key = fields.CharField(max_length=30, null=True, description="当前阶段门")
    owner_id = fields.IntField(null=True, description="负责人ID")
    owner_name = fields.CharField(max_length=100, null=True, description="负责人姓名")
    priority = fields.CharField(max_length=20, default="normal", description="优先级")
    planned_start_date = fields.DateField(null=True, description="计划开始日期")
    planned_end_date = fields.DateField(null=True, description="计划结束日期")
    actual_start_date = fields.DateField(null=True, description="实际开始日期")
    actual_end_date = fields.DateField(null=True, description="实际结束日期")
    notes = fields.TextField(null=True, description="备注")
    deleted_at = fields.DatetimeField(null=True, description="删除时间")

    class Meta:
        table = "apps_kuaiplm_rd_projects"
        table_description = "快研发 - 研发项目"
        indexes = [
            ("tenant_id",),
            ("project_code",),
            ("status",),
            ("tenant_id", "project_type"),
            ("source_project_id",),
            ("gate_template_id",),
        ]

    class PydanticMeta:
        exclude = ["deleted_at"]


class RdProjectGate(BaseModel):
    """NPI 阶段门"""

    tenant_id = fields.IntField(description="租户ID")
    project_id = fields.IntField(description="项目ID")
    gate_key = fields.CharField(max_length=30, description="阶段门标识")
    gate_name = fields.CharField(max_length=100, description="阶段门名称")
    sort_order = fields.IntField(default=0, description="排序")
    status = fields.CharField(
        max_length=30,
        default=RdGateStatus.PENDING.value,
        description="阶段门状态",
    )
    planned_date = fields.DateField(null=True, description="计划日期")
    actual_date = fields.DateField(null=True, description="实际日期")
    reviewer_id = fields.IntField(null=True, description="评审人ID")
    reviewer_name = fields.CharField(max_length=100, null=True, description="评审人姓名")
    review_notes = fields.TextField(null=True, description="评审备注")
    criteria = fields.TextField(null=True, description="通过准则")
    milestone_role = fields.CharField(
        max_length=30,
        default=GateMilestoneRole.NONE.value,
        null=True,
        description="里程碑角色 none | spawn_delivery",
    )

    class Meta:
        table = "apps_kuaiplm_rd_project_gates"
        table_description = "快研发 - NPI 阶段门"
        unique_together = (("tenant_id", "project_id", "gate_key"),)
        indexes = [("tenant_id", "project_id")]


class RdProjectTask(BaseModel):
    """研发项目任务"""

    tenant_id = fields.IntField(description="租户ID")
    project_id = fields.IntField(description="项目ID")
    gate_id = fields.IntField(null=True, description="阶段门ID")
    parent_task_id = fields.IntField(null=True, description="父任务ID（仅支持一级子任务）")
    task_name = fields.CharField(max_length=200, description="任务名称")
    description = fields.TextField(null=True, description="任务描述")
    status = fields.CharField(
        max_length=30,
        default=RdTaskStatus.TODO.value,
        description="任务状态",
    )
    assignee_id = fields.IntField(null=True, description="执行人ID")
    assignee_name = fields.CharField(max_length=100, null=True, description="执行人姓名")
    due_date = fields.DateField(null=True, description="截止日期")
    completed_at = fields.DatetimeField(null=True, description="完成时间")
    sort_order = fields.IntField(default=0, description="排序")
    priority = fields.CharField(max_length=20, default="normal", description="优先级")
    deleted_at = fields.DatetimeField(null=True)

    class Meta:
        table = "apps_kuaiplm_rd_project_tasks"
        table_description = "快研发 - 项目任务"
        indexes = [("tenant_id", "project_id"), ("parent_task_id",)]

    class PydanticMeta:
        exclude = ["deleted_at"]


class RdProjectDeliverable(BaseModel):
    """研发项目交付物"""

    tenant_id = fields.IntField(description="租户ID")
    project_id = fields.IntField(description="项目ID")
    gate_id = fields.IntField(null=True, description="阶段门ID")
    name = fields.CharField(max_length=200, description="交付物名称")
    description = fields.TextField(null=True, description="描述")
    deliverable_type = fields.CharField(max_length=50, null=True, description="交付物类型")
    status = fields.CharField(
        max_length=30,
        default=RdDeliverableStatus.PENDING.value,
        description="状态",
    )
    file_url = fields.CharField(max_length=500, null=True, description="文件URL")
    file_name = fields.CharField(max_length=200, null=True, description="文件名")
    submitted_at = fields.DatetimeField(null=True, description="提交时间")
    approved_at = fields.DatetimeField(null=True, description="批准时间")
    deleted_at = fields.DatetimeField(null=True)

    class Meta:
        table = "apps_kuaiplm_rd_project_deliverables"
        table_description = "快研发 - 项目交付物"
        indexes = [("tenant_id", "project_id")]

    class PydanticMeta:
        exclude = ["deleted_at"]


class RdProjectLink(BaseModel):
    """研发项目关联（BOM/工艺/图纸/工单等）"""

    tenant_id = fields.IntField(description="租户ID")
    project_id = fields.IntField(description="项目ID")
    link_type = fields.CharField(
        max_length=50,
        default=RdProjectLinkType.OTHER.value,
        description="关联类型",
    )
    target_type = fields.CharField(max_length=50, description="目标类型")
    target_id = fields.IntField(null=True, description="目标ID")
    target_uuid = fields.CharField(max_length=36, null=True, description="目标UUID")
    target_code = fields.CharField(max_length=100, null=True, description="目标编码")
    target_name = fields.CharField(max_length=200, null=True, description="目标名称")
    notes = fields.TextField(null=True, description="备注")

    class Meta:
        table = "apps_kuaiplm_rd_project_links"
        table_description = "快研发 - 项目关联"
        indexes = [
            ("tenant_id", "project_id"),
            ("target_type", "target_id"),
        ]
